docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,173 @@
+--------------
+
+The handling of bear dependencies by the old core wasn't effective. The
+old core used a queuing mechanism to communicate between bear runs which
+led to problems of a GlobalBear not running if any one of its
+dependencies (a local bear) terminated. The NextGen-Core does away with
+that by introducing a ``DependencyBear``.
+
+This bear serves as a base class which parallelizes tasks for each dependency
+result. A bear dependent on other bears can specify its dependencies in
+``BEAR_DEPS``. For example, there are two bears ``Foo`` and ``Bar``
+and bear ``Bar`` depends on ``Foo``. This can be written as
+
+::
+
+    class BarBear(DependencyBear):
+        BEAR_DEPS = {FooBear}
+
+Multiple bears can be included as a dependency of a bear in the BEAR_DEPS
+field. The results of the dependency bears are saved in a dictionary
+called ``_dependency_results`` which is initialized in the ``__init()__``
+method of the class ``Bear.py`` and can be accessed using the method
+``dependency_results()`` also belonging to the same class.
+
+Writing A Dependency Bear
+-------------------------
+
+Let's consider a bear to be dependent on a project bear ``Fizz`` and a file bear
+``Buzz`` then the corresponding DependencyBear let's call it ``FizzBuzz`` will
+look like the following:
+
+::
+
+    class FizzBuzzBear(DependencyBear):
+        BEAR_DEPS = {FizzBear, BuzzBear}
+
+        def analyze(self, dependency_bear, dependency_result, a_number=100):
+            yield '{} ({}) - {}'.format(
+                dependency_bear.name, a_number, dependency_result)
+
+Usage of instance-bound ``Bear.BEAR_DEPS``
+------------------------------------------
+
+A bear might depend on multiple bears (to run) before its execution can begin.
+``Bear.BEAR_DEPS`` is just a collection (a set) of classes (Bear classes) that
+need to be executed before that bear can run. Once all these dependencies
+(Bears) have run their results are passed into ``self.dependency_results``.
+The results are in the form of a dictionary with the types of the bears and
+their corresponding results (in the form of a list) as *Key-Value* pairs.
+From the previous example if we try to access the BEAR_DEPS of the ``BarBear``
+we will get the result ``{<class 'coalib.core.Bear.FooBear'>}``.
+
+In the `__init__()` method of the class ``Bear.py`` the dependencies specified
+in the ``BEAR_DEPS`` are copied to every instance of a Bear run using
+``self.BEAR_DEPS = set(self.BEAR_DEPS)`` which makes runtime modifications
+are possible.
+
+Override bears and define task scheduling
+-----------------------------------------
+
+A nextgen bear has to have the following functions to perform analysis
+- ``generate_tasks``: This method is a part of the parent ``Bear.py`` class
+and returns tuples containing the positional arguments as a tuple and the
+keyword arguments in the form of a dict. These are actually the task objects
+that are scheduled and executed by the core. An absence of this method raises
```

Comment:
You should definitely show some examples, `ProjectBear.generate_tasks` or `FileBear.generate_tasks` look pretty weird, but it's actually quite simple (the data just has to be arranged correctly)